Feature/reafactor pooll2tx types

arnaubennassar opened the pull request:
<!-- 🎉 Thank you for the PR!!! 🎉 -->

Partialy implements #833

### What does this PR does?

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->
- With this refactor, sending pool txs from an external Go project is as easy as marshaling `common.PoolL2Tx` into a JSON 
- Merge structs Merge `api.receivedPoolTx`, `api.testPoolTxSend` and `l2db.PoolL2TxWrite` into `common.PoolL2Tx`
- Use the same method to insert txs into the pool for testing and API (production)
- Prevent method `func (tx *PoolL2Tx) RqTxCompressedDataV2() (*big.Int, error)` from modifying `tx.RqAmount` value 

Note that `common.formats.go` and `common.formats.go` (which represents about 70% of the added code) are basically copied code from `api/apitypes` and `api` packages. In future PRs I will finish the refactoring by deleting the equivalent code from `./api/**`. I didn't make it in this one since the amount of altered code would have made the PR very hard to reason about.

### How to test?

<!-- What steps in order should someone run to test -->
Since this is a refactor, existing tests should be valid

## Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Include tests
- [x] Respect code style and lint
- [ ] Update swagger file (if needed)
- [ ] Update documentation (*.md) (if needed)

Mikelle commented on common/pooll2tx.go:
```diff
@@ -410,3 +422,165 @@ const (
+		return true
+	default:
+		return false
+	}
```
```suggestion
switch s {	
    case PoolL2TxStatePending, PoolL2TxStateForging, PoolL2TxStateForged, PoolL2TxStateInvalid: 
        return true	
    default: 
      return false	
 }
```

Mikelle commented on db/l2db/l2db.go:
```diff
@@ -147,58 +149,107 @@ func (l2db *L2DB) UpdateTxsInfo(txs []common.PoolL2Tx, batchNum common.BatchNum)
+	// if tx.ToIdx != 0 {
+	// 	toIdx = new(common.Idx)
+	// 	*toIdx = tx.ToIdx
+	// }
```
Is this commented code necessary?

Mikelle commented on db/l2db/l2db.go:
```diff
@@ -147,58 +149,107 @@ func (l2db *L2DB) UpdateTxsInfo(txs []common.PoolL2Tx, batchNum common.BatchNum)
-		insertTx.ToBJJ = &tx.ToBJJ
+		// RqFromIdx
+		rqFromIdx = new(common.Idx)
+		*rqFromIdx = tx.RqFromIdx
```
```suggestion
		rqFromIdx = &tx.RqFromIdx
```
Maybe is it possible to create it like this and don't create a redundant assignment? Applicable to all the following assignments after this.

Mikelle commented on db/l2db/l2db.go:
```diff
@@ -147,58 +149,107 @@ func (l2db *L2DB) UpdateTxsInfo(txs []common.PoolL2Tx, batchNum common.BatchNum)
+func (l2db *L2DB) addTx(tx *common.PoolL2Tx, checkPoolIsFull bool) (sql.Result, error) {
+	// Prepare extra DB fields and nullables
+	// AmountFloat
+	f := new(big.Float).SetInt((*big.Int)(tx.Amount))
```
Isn't tx.Amount already have *big.Int type?

arnaubennassar replied:
yes, this is used to store the amount as float in the DB. There is a column in the table `amount_f` that requires it

Mikelle commented on db/l2db/l2db.go:
```diff
@@ -147,58 +149,107 @@ func (l2db *L2DB) UpdateTxsInfo(txs []common.PoolL2Tx, batchNum common.BatchNum)
+	var rqTokenID *common.TokenID
+	var rqAmount *string
+	var rqFee *common.FeeSelector
+	var rqNonce *common.Nonce
```
```suggestion
	var (
	   info, rqAmount *string
	   rqFromIdx, rqToIdx *common.Idx 
	   rqToEthAddr *ethCommon.Address 
	   rqToBJJ *babyjub.PublicKeyComp 
	   rqTokenID *common.TokenID
	   rqFee *common.FeeSelector
	   rqNonce *common.Nonce
	   )
```

Mikelle commented on common/pooll2tx.go:
```diff
@@ -410,3 +422,165 @@ const (
+	if receivedJSON.State != "" {
+		state = PoolL2TxState(receivedJSON.State)
+		if !state.IsValid() {
+			return errors.New("Invalid state")
```
Maybe this way log message be better?
```suggestion
			return fmt.Errorf("Invalid state: %s", state)
```

Mikelle commented on common/formats.go:
```diff
@@ -0,0 +1,309 @@
+		// src is []byte
+		*e = *(NewEthSignature(srcBytes))
+		return nil
+	} else {
```
This else seems not needed, bcs you have returns in previous if cases

Mikelle commented on common/formats.go:
```diff
@@ -0,0 +1,309 @@
+
+func hezStrToBJJ(s string) (babyjub.PublicKeyComp, error) {
+	const decodedLen = 33
+	const encodedLen = 44
```
```suggestion
const (
	decodedLen = 33
	encodedLen = 44
)
```

arnaubennassar commented on common/pooll2tx.go:
```diff
@@ -410,3 +422,165 @@ const (
+		return true
+	case PoolL2TxStateForged:
+		return true
+	case PoolL2TxStateInvalid:
```
This state indicates that the tx is invalid, but it's one of the supported states therefore the state type is valid. A bit `meta` xD

arnaubennassar commented:
@Mikelle @ARR552 I know that you already approved this PR, but instead of going to `v1.4.0` it will go to `v2.0.0`. For this reason I've rebased the branch to `feature/atomic-tx`, invalidating your review. However since you already had a good look at the original code, let me help you by pointing the actual changes that I've made since your review:

- In `api/txspool_test.go`:
    - Added support for RqTxID
    - use `for i := range poolTxs` instead of `for _, tx := range poolTxs` to avoid problems with pointers
    - use common.PoolL2Tx instead of the test struct for the atomic related tests
- In `api/txspool.go`:
    - Change `l2db.PoolL2TxWrite` ==> `common.PoolL2T` for atomic endpoint related
- In `common/pooll2tx.go`:
    - added `RqTxID` to marshaler and unmarshaler
- In `db/l2db/apiqueries.go`:
    - In `func (l2db *L2DB) AddAtomicTxsAPI(txs []common.PoolL2Tx) error`:
        - changed how the AmountFloat is calculated as it's not longer part of the input struct
        - use the function `l2db.addTxs` to perform the insert in a consistent maner with other parts of the code
    - In `func (l2db *L2DB) AddTxAPI(tx *common.PoolL2Tx) error`:
        - adapt to the new impl of `l2db.addTxs`
- In `db/l2db/l2db.go`:
    - New implementation of `func (l2db *L2DB) addTxs(txs []common.PoolL2Tx, checkPoolIsFull bool) error` to support bulk insert (before was only possible to go one by one)

mfcastellani commented on api/txspool.go:
```diff
@@ -188,190 +180,68 @@ func (a *API) getPoolTxs(c *gin.Context) {
 	// Check destinatary, note that transactions that are not transfers
 	// will always be valid in terms of destinatary (they use special ToIdx by protocol)
-	switch poolTx.Type {
+	switch tx.Type {
```
One question, if`tx.Type` was not set this can be a problem. Is possible to have `tx` without `Type` set?

arnaubennassar replied:
Not in this part of the code: a few lines above this one we're calling `_, err := common.NewPoolL2Tx(&tx)`, which sets the `Type` if it wasn't, and returns an error if it had a wrong value

mfcastellani commented:
Testing here at my machine I got some fails when run `make test`. Is possible to be something related to my environment:

```
=== RUN   TestTxSelectorBatchBuilderZKInputsMinimumFlow0
2021-06-04T15:08:30-03:00	INFO	db/utils.go:61	successfully ran 0 migrations Up
2021-06-04T15:08:30-03:00	INFO	db/utils.go:77	successfully ran 4 migrations Down
2021-06-04T15:08:31-03:00	INFO	db/utils.go:61	successfully ran 4 migrations Up
2021-06-04T15:08:31-03:00	DEBUG	statedb/statedb.go:524	Making StateDB ResetFromSynchronizer	{"batch": 0, "type": "batchbuilder"}
2021-06-04T15:08:31-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:1
2021-06-04T15:08:31-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 0, "l1UserFutureTxs": 0}
2021-06-04T15:08:31-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:31-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 0}
2021-06-04T15:08:31-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 1, "type": "txselector"}
2021-06-04T15:08:31-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 1, "type": "batchbuilder"}
2021-06-04T15:08:31-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:31-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:2
2021-06-04T15:08:31-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 0, "l1UserFutureTxs": 0}
2021-06-04T15:08:31-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:31-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 0}
2021-06-04T15:08:31-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 2, "type": "txselector"}
2021-06-04T15:08:32-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 2, "type": "batchbuilder"}
2021-06-04T15:08:32-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:32-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:3
2021-06-04T15:08:32-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 2, "l1UserFutureTxs": 0}
2021-06-04T15:08:32-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00e979da4b80d60a17ce56fa19278c6f3a7e1b43359fb8a8ea46d0264de7d653ab"}
2021-06-04T15:08:32-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00af9bf96eb60f2d618519402a2f6b07057a034fa2baefd379fe8e1c969f1c5cf4"}
2021-06-04T15:08:32-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:32-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 2}
2021-06-04T15:08:32-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 3, "type": "txselector"}
2021-06-04T15:08:32-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 3, "type": "batchbuilder"}
2021-06-04T15:08:33-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:33-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:4
2021-06-04T15:08:33-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 1, "l1UserFutureTxs": 0}
2021-06-04T15:08:33-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00057b74a773eb9a6eac8deb118ff743b8add8f7fd8843f46f934ddc9868bd3b90"}
2021-06-04T15:08:33-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:33-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 1}
2021-06-04T15:08:33-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 4, "type": "txselector"}
2021-06-04T15:08:33-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 4, "type": "batchbuilder"}
2021-06-04T15:08:33-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:33-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:5
2021-06-04T15:08:33-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 0, "l1UserFutureTxs": 0}
2021-06-04T15:08:33-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:33-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 0}
2021-06-04T15:08:33-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 5, "type": "txselector"}
2021-06-04T15:08:33-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 5, "type": "batchbuilder"}
2021-06-04T15:08:34-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:34-03:00	DEBUG	zkproof/flows_test.go:153	block:0 batch:6
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 1, "l1UserFutureTxs": 0}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x005d7ee63390a6c95f23a97607f0b66edd44b7999f9a14a16d92107c02baee0700"}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:34-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 1}
2021-06-04T15:08:34-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 6, "type": "txselector"}
2021-06-04T15:08:34-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 6, "type": "batchbuilder"}
2021-06-04T15:08:34-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:34-03:00	DEBUG	zkproof/flows_test.go:169	block:0 batch:7
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 1, "l1UserFutureTxs": 0}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00930696d03ae0a1e6150b6ccb88043cb539a4e06a7f8baf213029ce9a0600197e"}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 2
2021-06-04T15:08:34-03:00	DEBUG	txselector/txgroup.go:157	TxSelector: new coordinator L1 tx CreateAccountDeposit	{"Addr": "0x1efF47bc3a10a45D4B230B5d10E37751FE6AA718", "BJJ": "7a515b54e07eb315df3b27601e616a7640500b4b5e1896b6c5b8c371d8d0051d", "TokenID": 0}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txgroup.go:196	TxSelector: new user L1 TransferToEthAddr tx CreateAccountDeposit	{"Addr": "0x6813Eb9362372EEF6200f3b1dbC3f819671cBA69", "BJJ": "ddeb5ed1d187bf9fbc0348ead28e476c4f1552fc7d144b896d75af5557fc652d", "TokenID": 0}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txgroup.go:232	TxSelector: processing new L1 tx	{"TxID": "0x000000000000000000000000000000000000000000000000000000000000000000"}
2021-06-04T15:08:34-03:00	DEBUG	txselector/txgroup.go:232	TxSelector: processing new L1 tx	{"TxID": "0x000000000000000000000000000000000000000000000000000000000000000000"}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "259", "txs": 1}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txgroup.go:157	TxSelector: new coordinator L1 tx CreateAccountDeposit	{"Addr": "0x1efF47bc3a10a45D4B230B5d10E37751FE6AA718", "BJJ": "7a515b54e07eb315df3b27601e616a7640500b4b5e1896b6c5b8c371d8d0051d", "TokenID": 1}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txgroup.go:196	TxSelector: new user L1 TransferToEthAddr tx CreateAccountDeposit	{"Addr": "0x2B5AD5c4795c026514f8317c7a215E218DcCD6cF", "BJJ": "50ac21cece6c195ae12f2e67001184b72969e509656643735d7961a904871b30", "TokenID": 1}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txgroup.go:232	TxSelector: processing new L1 tx	{"TxID": "0x000000000000000000000000000000000000000000000000000000000000000000"}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txgroup.go:232	TxSelector: processing new L1 tx	{"TxID": "0x000000000000000000000000000000000000000000000000000000000000000000"}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "258", "txs": 1}
2021-06-04T15:08:35-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 7}
2021-06-04T15:08:35-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 7, "type": "txselector"}
2021-06-04T15:08:36-03:00	DEBUG	txprocessor/txprocessor.go:1059	No coord Idx to receive fee	{"tx": {"IsL1":false,"TxID":"0x0222074744920dfdae4e31bb893074abcf3feccf50642831bac0dbdd657d45205f","Type":"TransferToEthAddr","Position":0,"FromIdx":258,"ToIdx":0,"Amount":200,"AmountFloat":0,"TokenID":1,"USD":null,"BatchNum":null,"EthBlockNum":0,"ToForgeL1TxsNum":null,"UserOrigin":null,"FromEthAddr":"0x0000000000000000000000000000000000000000","FromBJJ":"0000000000000000000000000000000000000000000000000000000000000000","DepositAmount":null,"DepositAmountFloat":null,"DepositAmountUSD":null,"Fee":126,"FeeUSD":null,"Nonce":0}}
2021-06-04T15:08:36-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 7, "type": "batchbuilder"}
    flows_test.go:189:
        	Error Trace:	flows_test.go:189
        	Error:      	Not equal:
        	            	expected: "17845026419825606568297898952886451490968178996859596985115147673317944565005"
        	            	actual  : "4103734334789947367650595587308238294442099148900595478318607388961717640220"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-17845026419825606568297898952886451490968178996859596985115147673317944565005
        	            	+4103734334789947367650595587308238294442099148900595478318607388961717640220
        	Test:       	TestTxSelectorBatchBuilderZKInputsMinimumFlow0
2021-06-04T15:08:36-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:36-03:00	DEBUG	zkproof/flows_test.go:200	block:0 batch:8
2021-06-04T15:08:36-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 0, "l1UserFutureTxs": 0}
2021-06-04T15:08:36-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 4
2021-06-04T15:08:36-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "256", "txs": 2}
2021-06-04T15:08:36-03:00	DEBUG	txprocessor/txprocessor.go:1059	No coord Idx to receive fee	{"tx": {"IsL1":false,"TxID":"0x0209660ee0c15461ad96f9baac8213875ab8dff74b9b8e04ee4f31ddaf2eea36f5","Type":"Transfer","Position":0,"FromIdx":264,"ToIdx":256,"Amount":50,"AmountFloat":0,"TokenID":0,"USD":null,"BatchNum":null,"EthBlockNum":0,"ToForgeL1TxsNum":null,"UserOrigin":null,"FromEthAddr":"0x0000000000000000000000000000000000000000","FromBJJ":"0000000000000000000000000000000000000000000000000000000000000000","DepositAmount":null,"DepositAmountFloat":null,"DepositAmountUSD":null,"Fee":126,"FeeUSD":null,"Nonce":0}}
2021-06-04T15:08:36-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "264", "txs": 1}
2021-06-04T15:08:36-03:00	DEBUG	txselector/txgroup.go:344	TxSelector: insufficient funds	{"TxID": "0x02866abb32759f0f85542f53be319ef763764aa9faadce7353628f135ddd9c5f2e"}
2021-06-04T15:08:36-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "262", "txs": 1}
2021-06-04T15:08:36-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 3}
2021-06-04T15:08:36-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 8, "type": "txselector"}
2021-06-04T15:08:37-03:00	DEBUG	txprocessor/txprocessor.go:1059	No coord Idx to receive fee	{"tx": {"IsL1":false,"TxID":"0x0209660ee0c15461ad96f9baac8213875ab8dff74b9b8e04ee4f31ddaf2eea36f5","Type":"Transfer","Position":0,"FromIdx":264,"ToIdx":256,"Amount":50,"AmountFloat":0,"TokenID":0,"USD":null,"BatchNum":null,"EthBlockNum":0,"ToForgeL1TxsNum":null,"UserOrigin":null,"FromEthAddr":"0x0000000000000000000000000000000000000000","FromBJJ":"0000000000000000000000000000000000000000000000000000000000000000","DepositAmount":null,"DepositAmountFloat":null,"DepositAmountUSD":null,"Fee":126,"FeeUSD":null,"Nonce":0}}
2021-06-04T15:08:37-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 8, "type": "batchbuilder"}
    flows_test.go:219:
        	Error Trace:	flows_test.go:219
        	Error:      	Not equal:
        	            	expected: "5011486595982139468087543085074814351795839959948110063153096719520312696275"
        	            	actual  : "3999127613851587618905669990719687190398036089498713875270929682195517115654"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-5011486595982139468087543085074814351795839959948110063153096719520312696275
        	            	+3999127613851587618905669990719687190398036089498713875270929682195517115654
        	Test:       	TestTxSelectorBatchBuilderZKInputsMinimumFlow0
2021-06-04T15:08:37-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:37-03:00	DEBUG	zkproof/flows_test.go:229	(batch9) block:1 batch:1
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 4, "l1UserFutureTxs": 0}
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00446bdc27a508f491c1c4c0f1f2026ca3958639bc3111f6d5355a358fd6c83c2b"}
2021-06-04T15:08:37-03:00	DEBUG	txprocessor/txprocessor.go:1443	EffectiveAmount & EffectiveDepositAmount = 0: tx.TokenID (0) !=sender account TokenID (1)
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00caef6d804c6bbd96191b2dd33a08b44ad719d39db1c6e08adcc837e3d439e4e7"}
2021-06-04T15:08:37-03:00	DEBUG	txprocessor/txprocessor.go:1443	EffectiveAmount & EffectiveDepositAmount = 0: tx.TokenID (0) !=sender account TokenID (1)
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x00911df78bc06fc9e0ca68d63f20c86a0fc6374d6a9de35e7b1757dd9c47ef6c55"}
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:123	TxSelector: processing L1 user tx	{"TxID": "0x0030926e7bcb8ddfdc2a9ed6a2de3f51e0fd09c2ae6e62fad9e781c495047c5f57"}
2021-06-04T15:08:37-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 2
2021-06-04T15:08:37-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "260", "txs": 1}
2021-06-04T15:08:37-03:00	DEBUG	txselector/txbatch.go:207	TxSelector: tx group created	{"group": "259", "txs": 1}
2021-06-04T15:08:37-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 6}
2021-06-04T15:08:37-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 9, "type": "txselector"}
2021-06-04T15:08:38-03:00	DEBUG	txprocessor/txprocessor.go:1443	EffectiveAmount & EffectiveDepositAmount = 0: tx.TokenID (0) !=sender account TokenID (1)
2021-06-04T15:08:38-03:00	DEBUG	txprocessor/txprocessor.go:1443	EffectiveAmount & EffectiveDepositAmount = 0: tx.TokenID (0) !=sender account TokenID (1)
2021-06-04T15:08:38-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 9, "type": "batchbuilder"}
    flows_test.go:246:
        	Error Trace:	flows_test.go:246
        	Error:      	Not equal:
        	            	expected: "6530394748986447282151711357194248066562667891309466162814480602176851300695"
        	            	actual  : "18834200399613399594019274526889670645661725251230742025078949033688445361340"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-6530394748986447282151711357194248066562667891309466162814480602176851300695
        	            	+18834200399613399594019274526889670645661725251230742025078949033688445361340
        	Test:       	TestTxSelectorBatchBuilderZKInputsMinimumFlow0
2021-06-04T15:08:38-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
2021-06-04T15:08:38-03:00	DEBUG	zkproof/flows_test.go:256	(batch10) block:1 batch:2
2021-06-04T15:08:38-03:00	DEBUG	txselector/txselector.go:99	TxSelector: GetL1L2TxSelection	{"l1UserTxs": 0, "l1UserFutureTxs": 0}
2021-06-04T15:08:38-03:00	DEBUG	txselector/txselector.go:135	TxSelector: pending txs: 0
2021-06-04T15:08:38-03:00	DEBUG	txselector/txbatch.go:143	TxSelector: batch pruned	{"txs": 0}
2021-06-04T15:08:38-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 10, "type": "txselector"}
2021-06-04T15:08:39-03:00	DEBUG	statedb/statedb.go:226	Making StateDB checkpoint	{"batch": 10, "type": "batchbuilder"}
    flows_test.go:268:
        	Error Trace:	flows_test.go:268
        	Error:      	Not equal:
        	            	expected: "6530394748986447282151711357194248066562667891309466162814480602176851300695"
        	            	actual  : "18834200399613399594019274526889670645661725251230742025078949033688445361340"

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-6530394748986447282151711357194248066562667891309466162814480602176851300695
        	            	+18834200399613399594019274526889670645661725251230742025078949033688445361340
        	Test:       	TestTxSelectorBatchBuilderZKInputsMinimumFlow0
2021-06-04T15:08:39-03:00	DEBUG	zkproof/zkproof_test.go:64	No PROOF_SERVER_URL defined, not using ProofServer
--- FAIL: TestTxSelectorBatchBuilderZKInputsMinimumFlow0 (9.48s)
FAIL
FAIL	github.com/hermeznetwork/hermez-node/test/zkproof	10.866s
```

arnaubennassar commented:
> Testing here at my machine I got some fails when run make test. Is possible to be something related to my environment

Yes, apparently the test fails like 5% of the times (on my machine). This test is the one of the compatibility canary we're fixing in #857 . 

Apart from that, I believe this weird behavior is not caused by the changes of this PR, as the data that causes the test to fail should be exactly the same as before this PR.

I believe we should merge this and investigate further this problem, and what is causing this
